### PR TITLE
RUST-870 Deserialize server response directly from raw BSON bytes

### PR DESCRIPTION
**Pull request by patrickfreed**
RUST-870

This PR updates the driver to leverage the new `bson::from_slice` serde functionality in the BSON library to deserialize all server response directly from BSON rather than going through `Document` first, resulting in significant performance improvements.

### CODE REVIEW

File: src/cursor/session.rs [hunk 1]
```diff
@@ -55,15 +55,15 @@ where
     exhausted: bool,
     client: Client,
     info: CursorInformation,
-    buffer: VecDeque<Document>,
+    buffer: VecDeque<T>,
     _phantom: std::marker::PhantomData<T>,
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply src/cursor/session.rs]: nope, good catch, removed.

File: src/client/executor.rs [hunk 2]
```diff
@@ -393,12 +450,16 @@ impl Client {
             }
             Ok(response) => {
                 self.emit_command_event(|handler| {
+                    let response_doc: Document = response
+                        .raw
+                        .body()
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply src/client/executor.rs]: Yep that's right, and that's a good idea, added a line about this to the `command_event_handler` option in `ClientOptions`.

File: src/client/executor.rs [hunk 3]
```diff
@@ -393,12 +450,16 @@ impl Client {
             }
             Ok(response) => {
                 self.emit_command_event(|handler| {
+                    let response_doc: Document = response
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply src/client/executor.rs]: fixed

File: Cargo.toml [hunk 4]
```diff
@@ -38,7 +38,7 @@ bson-uuid-0_8 = ["bson/uuid-0_8"]
 async-trait = "0.1.42"
 base64 = "0.13.0"
 bitflags = "1.1.0"
-bson = "2.0.0-beta.2"
+bson = { git = "https://github.com/mongodb/bson-rust" }
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply Cargo.toml:41]: Yep, filed RUST-901 to remind us to do that.

**Review by abr-egn — APPROVED** [review]

**Review by isabelatkinson — APPROVED** [review]

**Comment by patrickfreed** [comment]
Re-requesting review (sorry!) for the snapshot time / `atClusterTime` changes needed after rebasing off of master, should only need to take a look at the last two commits.

**Review by abr-egn — APPROVED** [review]
LGTM - thanks for handling the merge!

**Review by isabelatkinson — APPROVED** [review]

**Pull request merged**